lib/operations/mongo_client_ops.js

```diff
@@ -600,4 +628,80 @@ function validOptions(options) {
   }
 }
 
+const apmLoggingEvents = {
+  command: ['commandStarted', 'commandSucceeded', 'commandFailed'],
+  sdam: [
+    'serverOpening',
+    'serverDescriptionChanged',
+    'serverHeartbeatStarted',
+    'serverHeartbeatSucceeded',
+    'serverHeartbeatFailed',
+    'serverClosed',
+    'topologyOpening',
+    'topologyClosed',
+    'topologyDescriptionChanged'
+  ]
+};
+
+function apmLoggingEventHandler(client) {
+  return function(err) {
+    if (!err) return;
+    if (!client.topology) throw err;
+    client.close(true, () => {
+      throw err;
+    });
+  };
+}
+
+function getMonitoringCategories(monitor) {
+  const allMonitors = ['command', 'sdam'];
+  // enable all monitors if monitor = all
+  let monitors = monitor === 'all' ? allMonitors : monitor;
+  // ensure `monitors` is an array
+  monitors = Array.isArray(monitors) ? monitors : [monitors];
+  return monitors;
+}
+
+function resolveMonitoringOutputFile(monitorOut) {
+  const defaultFileName = `${process.cwd()}/monitor.log`;
+  if (!monitorOut) return defaultFileName;
+  if (monitorOut === 'stderr' || monitorOut === 'stdout') return monitorOut;
+  return path.resolve(monitorOut);
+}
+
+function createMonitoringStream(fileName) {
+  if (fileName === 'stderr' || fileName === 'stdout') return process[fileName];
+  if (fs.existsSync(fileName)) {
```

Comment:
I'm sort of worried about these few lines for a few reasons:
- we should avoid synchronous code wherever possible
- if a user specifies a filename, and it exists, we just throw on extra metadata to the filename?

I think if a user has specified a name, and it exists, we should error out or append to the file. My guess is that our preference is actually that no filename is specified, and we just create a default constructed unique filename.

Reply:
I think erroring could be annoying to the user, because if they have a piece of code with a custom filename that they want to run repeatedly, they would have to manually rename the file each time to avoid the error. 

Appending could be good - should there be some sort of separator in the log file to indicate that a new batch of log messages was started?

Reply:
Generally what utilities do in this case is truncate the file, I'm okay with either approach but I don't think we should spend too much time on this. My guess is that the general use case here will be to provide no `monitorOut` _at all_, and just assume that there is a well named file in the folder I ran my sample code from named like `mongodb_monitoring-<date-thing>.log` or whatever